src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -219,6 +217,22 @@ public void clusterChanged(ClusterChangedEvent event) {
 
             // check and write changes in indices
             for (IndexMetaData indexMetaData : newMetaData) {
+
+                boolean shardsAllocatedOnThisNodeInLastClusterState = true;
+                if (isDataOnlyNode(state)) {
+                    boolean shardsCurrentlyAllocatedOnThisNode = shardsAllocatedOnLocalNode(state, indexMetaData);
+                    shardsAllocatedOnThisNodeInLastClusterState = shardsAllocatedOnLocalNode(event.previousState(), indexMetaData);
+
+                    if (shardsCurrentlyAllocatedOnThisNode == false) {
+                        // remove the index state for this index if it is only a data node
+                        // only delete if the last shard was removed
+                        if (shardsAllocatedOnThisNodeInLastClusterState) {
+                            removeIndexState(indexMetaData);
```

Comment:
ohh, I think I see what happens, in IndicesClusterStateService, we just remove the index and not delete it from the file system, which is good. In IndicesStore, only there we actually delete the shard content once the shard is allocated on all the other nodes. I think that we need to add a logic there that if the there are no more shards around, we need to also delete the index itself (which will delete its metadata).